File: app/src/main/java/com/raredev/theblocklogics/adapters/ProjectsAdapter.java
package com.raredev.theblocklogics.adapters;

import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.raredev.theblocklogics.databinding.LayoutProjectItemBinding;
import com.raredev.theblocklogics.models.Project;
import com.raredev.theblocklogics.utils.Constants;
import com.raredev.theblocklogics.viewmodel.MainViewModel;
import java.io.File;

public class ProjectsAdapter extends RecyclerView.Adapter<ProjectsAdapter.VH> {

  private LayoutInflater layoutInflater;
  private MainViewModel mainViewModel;

  private ProjectListener listener;

  public ProjectsAdapter(
      @NonNull LayoutInflater layoutInflater, @NonNull MainViewModel mainViewModel) {
    this.layoutInflater = layoutInflater;
    this.mainViewModel = mainViewModel;
  }

  @Override
  public VH onCreateViewHolder(ViewGroup parent, int viewType) {
    return new VH(LayoutProjectItemBinding.inflate(layoutInflater, parent, false));
  }

  @Override
  public void onBindViewHolder(VH holder, int position) {
    holder.itemView.setAnimation(
        AnimationUtils.loadAnimation(holder.itemView.getContext(), android.R.anim.fade_in));
    var project = mainViewModel.getProjectsList().get(position);

    holder.binding.icon.setImageDrawable(Drawable.createFromPath(project.getAppIconPath()));
    holder.binding.name.setText(project.getAppName());
    holder.binding.appPackage.setText(project.getAppPackage());

    holder.itemView.setOnClickListener(
        v -> {
          if (listener != null) listener.onProjectClick(project);
        });

    holder.binding.menu.setOnClickListener(
        v -> {
          if (listener != null) listener.onProjectMenuClick(v, project);
        });
  }

  @Override
  public int getItemCount() {
    return mainViewModel.getProjectsList().size();
  }

  public void setProjectListener(ProjectListener listener) {
    this.listener = listener;
  }

  public interface ProjectListener {
    void onProjectClick(Project project);

    void onProjectMenuClick(View v, Project project);
  }

  class VH extends RecyclerView.ViewHolder {
    LayoutProjectItemBinding binding;

    public VH(@NonNull LayoutProjectItemBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }
  }
}

File: app/src/main/java/com/raredev/theblocklogics/adapters/ProjectsAdapter.kt
package com.raredev.theblocklogics.adapters

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.recyclerview.widget.RecyclerView
import com.raredev.theblocklogics.databinding.LayoutProjectItemBinding
import com.raredev.theblocklogics.models.Project
import com.raredev.theblocklogics.utils.Constants
import com.raredev.theblocklogics.viewmodel.MainViewModel
import java.io.File

public class ProjectsAdapter(
  val layoutInflater: LayoutInflater,
  val viewModel: MainViewModel,
  val listener: ProjectListener
): RecyclerView.Adapter<ProjectsAdapter.VH>() {

  class VH(internal val binding: LayoutProjectItemBinding):
    RecyclerView.ViewHolder(binding.root)

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
    return VH(LayoutProjectItemBinding.inflate(layoutInflater, parent, false))
  }

  override fun onBindViewHolder(holder: VH, position: Int) {
    holder.binding.apply {
      root.setAnimation(
        AnimationUtils.loadAnimation(holder.itemView.getContext(), android.R.anim.fade_in))
      val projects = viewModel.projects.value
      projects?.let {
        val project = projects[position]

        icon.setImageDrawable(Drawable.createFromPath(project.getAppIconPath()))
        name.text = project.appName
        appPackage.text = project.appPackage

        root.setOnClickListener {
          listener.onProjectClick(project)
        }

        menu.setOnClickListener { v ->
          listener.onProjectMenuClick(v, project)
        }
      }
    }
  }

  override fun getItemCount(): Int {
    val projects = viewModel.projects.value
    return projects?.size ?: 0
  }

  public interface ProjectListener {
    fun onProjectClick(project: Project)

    fun onProjectMenuClick(v: View, project: Project)
  }
}